### ISSUE TITLE
Instantiate function not part of ActiveRecordInterface

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/yii2#9745]
ActiveQueryTrait calls the instantiate function on the $modelClass, but instantiate is not listed in the ActiveRecordInterface.

All functions that seem to be missing from the interface:
- instantiate
- populateRecord
- afterFind
- populateRelation

Code snippet from ActiveRelationTrait:273, checks for interface then calls functions not defined in the interface!

```
    if ($primaryModel instanceof ActiveRecordInterface) {
                $primaryModel->populateRelation($name, $value);
```

**Comment by SamMousa** [comment]
Also, _related and populateRelation implementations should be moved to the ActiveRelationTrait I think..

Not sure how you decide what is in a trait and what not, but I find myself copying the $_related and populateRelation function from BaseActiveRecord to my own class.

**Comment by cebe** [comment]
ActiveRelationTrait is part of the ActiveQuery, not ActiveRecord.

**Comment by SamMousa** [comment]
Oh thats my bad then. Hard to visualize all of it in my mind ;) still issue remains valid.

**Comment by lynicidn** [comment]
no. Populate relation and populate record it not query logic.
If i don't use ar and use data mapper.
I think it logic of Repository pattern - and now it activerecord

**Comment by lynicidn** [comment]
i see active record as https://github.com/paper-cube/active-data-mapper/blob/master/RepositoryInterface.php#L39 (it's very very simple code and WIP - idea rewrite ar logic to data mapper). Upper code it copy paste from 
BaseActiveRecord => BaseRepository + Entity
ActiveRecord => Repository + Entity
ActiveRecordInterface => RepositoryInterface + EntityInterface

**Comment by SamMousa** [comment]
I have no idea what you are talking about.

**Comment by lynicidn** [comment]
it bad idea move mapper logic to query

**Comment by SamMousa** [comment]
@lynicidn, I am sorry but I cannot make sense of your English.

**Comment by SamMousa** [comment]
Function getRelatedRecords in BaseActiveRecord is also missing from the ActiveRecordInterface interface.

**Comment by lynicidn** [comment]
ActiveRecord it common logic for repository, mapper, entity(model)
getRelatedRecords it inner logic, this method not ensure related records it getter for eager relations on init and lazy relations before findFor @see __get

**Comment by SamMousa** [comment]
@cebe : I'd like to make a PR where I update the ActiveRecordInterface with the following functions:

- populateRelation

There are some requirements for the $modelClass property of ActiveQueryTrait that are not captured in code anywhere:
The class $modelClass must implement the following functions:
- public static instantiate
- public static populateRecord
- public afterFind

Type hinting inside ActiveQueryTrait (line 109) suggests createModels returns ActiveRecord objects, but the type $modelClass is never checked, and should this type not be ActiveRecordInterface instead?

And if it should be ActiveRecordInterface, the 3 functions listed above should be added to that interface.

**Comment by SamMousa** [comment]
@cebe In #1282 you mention I can get a list of populated relations, but that function is not part of the ActiveRecordInterface, should it be added as well?

**Comment by cebe** [comment]
you mean `getRelatedRecords()` and `isRelationPopulated()`? may be candidates too.

**Comment by SamMousa** [comment]
!

**Comment by SamMousa** [comment]
Yes

**Comment by cebe** [comment]
Also considering to add abstract methods to ActiveRelationTrait like this:

```diff
diff --git a/framework/db/ActiveRelationTrait.php b/framework/db/ActiveRelationTrait.php
index 22f971f..4a97acb 100644
--- a/framework/db/ActiveRelationTrait.php
+++ b/framework/db/ActiveRelationTrait.php
@@ -17,9 +17,9 @@ use yii\base\InvalidParamException;
  * @author Carsten Brandt <mail@cebe.cc>
  * @since 2.0
  *
- * @method ActiveRecordInterface one()
- * @method ActiveRecordInterface[] all()
  * @property ActiveRecord $modelClass
+ * @property boolean $isArray
+ * @property string|null $indexBy
  */
 trait ActiveRelationTrait
 {
@@ -530,4 +530,28 @@ trait ActiveRelationTrait
 
         return $this->asArray()->all($primaryModel::getDb());
     }
+
+    /**
+     * @return ActiveQueryInterface
+     * @see ActiveQueryInterface::asArray()
+     */
+    abstract public function asArray($value = true);
+
+    /**
+     * @see ActiveQueryInterface::andWhere()
+     * @return ActiveQueryInterface
+     */
+    abstract public function andWhere($db = null);
+
+    /**
+     * @see ActiveQueryInterface::one()
+     * @return ActiveRecord|array|null
+     */
+    abstract public function one($db = null);
+
+    /**
+     * @see ActiveQueryInterface::all()
+     * @return array|ActiveRecordInterface[]
+     */
+    abstract public function all($db = null);
 }
```

**Comment by samdark** [comment]
Won't be implemented in 2.0.
